### PR TITLE
RWAHS-240 Initial import csv transformation mechanism

### PR DESCRIPTION
**Pull request by leftclickben**
As discussed, there are some leftovers to clean up, which are mostly marked with TODOs in the code.  I'll have to add tasks to clean this stuff up in the final sprint.

### CODE REVIEW

**Comment by leftclickben** [comment]
Just added a change (25a46b1) to map category names:
- Category names were causing an issue for importers (too long?).
- Also the category names were not very good for web display.
- This adds a map from source categories to target categories.
- Imports are working again now.

File: README.md [hunk 1]
```diff
@@ -18,6 +18,44 @@ MySQL and PHP are not prerequisites.  Although the eventual target system is Mag
 database, the actual target for the migration scripts is a Magento2-compatible CSV file, not the database.  Magento2
 itself requires PHP, and handles the task of importing from CSV into MySQL.
 
+## Setting Up
+
+Once this repo is cloned and prerequisites are installed, all that is required is to install the local node package
+dependencies using the standard command:
```
Comment (kehh) [inline README.md:24]: And installing the mdb-tools package.

Comment (leftclickben) [inline README.md:24]: You mean "and prerequisites are installed"?  (The previous heading being "Prerequisites")

Comment (leftclickben) [inline README.md:24]: ;-)

File: bin/process-images.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+                console.error('Could not read images path "%s", please check the path exists and try again', sourcePath);
+                return;
+            }
+            copyNextImage = function () {
+                var sourceImage;
+                if (sourceImages.length === 0) {
+                    return;
+                }
+                sourceImage = sourceImages.pop();
+                fse.copySync(sourcePath + '/' + sourceImage, targetPath + '/' + sourceImage.replace(/[^-\w.\/]+/g, '_'));
```
Comment (kehh) [inline bin/process-images.js:25]: Shouldn't this be `/[^-\w\.\/]+/g`?

File: bin/process-csv.js [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+                        .replace(/\s*[-(]\s*out\s+of\s+(?:stock|print|order).*$/i, '')
+                        .replace(/\s*[-(]\s*special\s+orders?\s+only.*$/i, '');
+
+                    imageFilename = _.find(imageFilenames, (filename) => filename.match(new RegExp('^' + record.BookCode + '_')));
+
+                    return {
+                        sku: 'BOOK-' + record.BookCode,
+                        store_view_code: '',
+                        attribute_set_code: 'Default',
+                        product_type: 'simple',
```
Comment (kehh) [inline bin/process-csv.js:61]: I'm wondering whether product type should be `book`? Does product type define what attributes it can have? (eg author)

File: bin/process-csv.js [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+            if (err) {
+                console.error('Could not read images path "%s", please check the path exists and try again', targetPath + '/images');
+                return;
+            }
+            fs.createReadStream(workPath + '/Books.csv')
+                .pipe(csv.parse({ columns: true }))
+                .pipe(csv.transform(function (record) {
+                    var title, imageFilename;
+
+                    if (!record.RRP) {
```
Comment (kehh) [inline bin/process-csv.js:46]: Do you skip many records using this rule? Should they just come through as draft records?

File: bin/process-images.js (hunk 2 above)
Comment (leftclickben) [inline bin/process-images.js:25]: Nah.  First empirically:

```
'foo bar\nbaz ^caret .dot'.replace(/[^-\w.\/]+/g, '_');
'foo_bar_baz_caret_.dot'
```

(The dot is allowed through as we want).

Outside of a character group, `.` means "match anything" which is actually a shorthand for a character group that includes every possible character.  Inside a character group, it only matches a `.` character.  In general, you only need to escape things in character groups that would otherwise have another meaning.  In the case of `.`, it never means "match anything" inside a character group, because "match anything" is itself a character group (and if you think about it, it's pretty meaningless to use "match anything" inside another character group because anything else you add will already be covered, which then makes the whole group construct also redundant).

To take a weirder example, the `^` character needs to be escaped or not depending on its _position_ in the group: `/[a^]/` and `/\^a]/` and `/[a\^]/` are all equivalent (match `a` or `^`), but `/[^a]/` is completely different (match anything except `a`):

```
> 'foo bar\nbaz ^caret .dot'.replace(/[\^a]/g, '_');
'foo b_r\nb_z _c_ret .dot'

> 'foo bar\nbaz ^caret .dot'.replace(/[a\^]/g, '_'); // escaping valid but redundant
'foo b_r\nb_z _c_ret .dot'

> 'foo bar\nbaz ^caret .dot'.replace(/[a^]/g, '_');
'foo b_r\nb_z _c_ret .dot'

> 'foo bar\nbaz ^caret .dot'.replace(/[^a]/g, '_');
'_____a___a____a________'
```

So, in summary, regular expressions are quite irregular at times.

File: bin/process-csv.js [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+                        custom_design: '',
+                        custom_design_from: '',
+                        custom_design_to: '',
+                        custom_layout_update: '',
+                        page_layout: '',
+                        product_options_container: '',
+                        msrp_display_actual_price_type: 'Use config',
+                        // TODO Can we assume this? Or should it be blank?
+                        country_of_manufacture: 'Australia',
+                        // TODO Derive correct value for `quantity_and_stock_status`
```
Comment (kehh) [inline bin/process-csv.js:108]: Doesn't this come from the publisher?

File: bin/process-csv.js (hunk 4 above)
Comment (leftclickben) [inline bin/process-csv.js:46]: Not sure.  I can put a TODO on it if you like.  It's a rule of the Magento importer that it has to have a price `> 0` or the import will fail (which it seems to really like doing).

File: bin/process-csv.js [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+                        use_config_min_sale_qty: 1,
+                        max_cart_qty: 0,
+                        use_config_max_sale_qty: 1,
+                        is_in_stock: Math.min(1, record.NoInStock),
+                        notify_on_stock_below: '',
+                        use_config_notify_stock_qty: 1,
+                        // TODO Find out what this does
+                        manage_stock: 1,
+                        use_config_manage_stock: 1,
+                        qty_increments: 0,
```
Comment (kehh) [inline bin/process-csv.js:128]: 1?

File: bin/process-csv.js (hunk 3 above)
Comment (leftclickben) [inline bin/process-csv.js:61]: This is a controlled vocabulary.  The `simple` product type is one without variants, isn't a bundle, etc.

Comment (leftclickben) [inline bin/process-csv.js:61]: It affects the way magento displays and processes the product, so it's effectively hardcoded into the framework.

File: bin/process-csv.js (hunk 6 above)
Comment (kehh) [inline bin/process-csv.js:128]: Sorry,  I see it's in config in the next line.

File: bin/process-csv.js (hunk 5 above)
Comment (leftclickben) [inline bin/process-csv.js:108]: "In stock" means that RWAHS has one or more copies, not that it's available from the publisher, so no.  

The source data has a field with the number in stock, so it's just a matter of deriving it, but it's left as a TODO for now.

File: bin/process-images.js (hunk 2 above)
Comment (kehh) [inline bin/process-images.js:25]: TIL

File: bin/process-csv.js (hunk 4 above)
Comment (kehh) [inline bin/process-csv.js:46]: Ok, please do. We can then at least give them a spreadsheet of skipped records.

File: bin/process-csv.js (hunk 5 above)
Comment (kehh) [inline bin/process-csv.js:108]: Oops I commented on the wrong line. I was talking about Country. But I don't think we can assume even that (many books have 'publisher X, London ' but printed in China,  so country of manufacture could easily not be Australia. But yes TODO caters for it.

File: bin/process-csv.js (hunk 4 above)
Comment (leftclickben) [inline bin/process-csv.js:46]: OK I've added a reporting mechanism for skipped records, on the line following this one (47) and from line 149 actually creates the `skipped-records.csv` file.  Apparently I can't attach CSV files to comments but here's the contents, it's only 12 records:

```
#,BookCode,Title,Reason for Skipping
1,545,Cossack Gold: The Romance of the Kimberleys  DO NOT USE,RRP is zero or missing
2,1201,"Journal of the Historical Society of South Australia - Number 41, 2013",RRP is zero or missing
3,1206,use this space,RRP is zero or missing
4,1228,use this space 1,RRP is zero or missing
5,1239,Vanishing Village,RRP is zero or missing
6,1256,use this space 3,RRP is zero or missing
7,1257,Banboo Curtain over Port Hedland,RRP is zero or missing
8,1300,Secret No Longer,RRP is zero or missing
9,1347,Old Fremantle Childhood,RRP is zero or missing
10,1429,C,RRP is zero or missing
11,1461,B,RRP is zero or missing
12,1463,Carala Zampatti,RRP is zero or missing
```

File: bin/process-csv.js (hunk 5 above)
Comment (leftclickben) [inline bin/process-csv.js:108]: Oh right that makes more sense :-)  Yeah it's not necessarily right for country, but they won't look at that much detail and if they do I'll show them the code and these comments.

File: bin/process-images.js (hunk 2 above)
Comment (leftclickben) [inline bin/process-images.js:25]: ...not to get Ben started on regular expressions ;-)
